README.md
# Notice: As of Aug 30, 2022, this wrapper is no longer updated or maintained.

# Chargify API wrapper for Ruby (using ActiveResource)

[![build status](https://secure.travis-ci.org/chargify/chargify_api_ares.png)](http://travis-ci.org/chargify/chargify_api_ares) [![Code Climate](https://codeclimate.com/github/chargify/chargify_api_ares/badges/gpa.svg)](https://codeclimate.com/github/chargify/chargify_api_ares)

This is a community-maintained Ruby wrapper for the [Chargify](https://www.chargify.com) API that leverages ActiveResource.

Though we do not proactively maintain it, Chargify is happy to review pull requests and manage the release process for the gem. We encourage community contributions to make the wrapper even more comprehensive and powerful.

If you notice an endpoint or method that is missing, please feel free to add it by submitting a pull request. We appreciate your support!

**Note:** we have bumped to v1.0.0 to indicate a backwards incompatible change to the responses from `Chargify::Subscription.charge` and `Chargify::Subscription.migrate`. Please see the HISTORY.md for more information.

**Please see important compatibility information at the bottom of this file.**

### Installation

This library can be installed as a gem. It is hosted on [Rubygems](http://rubygems.org).

You can install this library as a gem using the following command:

```bash
$ gem install chargify_api_ares
```

### Usage

Simply require this library before you use it:

```ruby
require 'chargify_api_ares'
```

If you're using Rails 3.x, you could include this gem in your Gemfile.

```ruby
gem 'chargify_api_ares'
```

Install the gem using the following command:
$ bundle install

If you're using Rails 2 you could include this gem in your configuration, i.e. in `environment.rb`

```ruby
config.gem 'chargify_api_ares'
```

To configure api key, site, domain, and protocol you can do the
following. Please note that this step is required.

#### Most common usage

```ruby
  Chargify.configure do |c|
    c.api_key   = "your_key_goes_here"
    c.subdomain = "test-site"
  end
```

#### Overriding the endpoint

```ruby
  Chargify.configure do |c|
    c.api_key   = "your_key_goes_here"
    c.site      = "https://test-site.some-domain.com"
  end
```

### Available configuration options

| Name      | Description                                                                                      | Default      | Required                   |
| --------- | ------------------------------------------------------------------------------------------------ | ------------ | -------------------------- |
| api_key   | The api_key credentials that are used to access your chargify account.                           | N/A          | Yes                        |
| subdomain | The subdomain (site name) of the chargify endpoint you are trying to interact with               | test         | Yes (unless site was used) |
| domain    | The domain of the endpoint, in which you want to interact with.                                  | chargify.com | No                         |
| protocol  | The endpoint protocol that you wish to use (http / https)                                        | https        | No                         |
| site      | This is meant to override all three of the previous settings eg: http://foo.bar.com              | N/A          | No                         |
| format    | The format of the request and response type that you want to deal with                           | xml          | No                         |
| timeout   | The time in seconds for a request to be valid. Will raise a timeout error if exceeds time limit. | N/A          | No                         |

Now you'll have access to classes the interact with the Chargify API, such as:

- `Chargify::Product`
- `Chargify::Customer`
- `Chargify::Subscription`

It allows you to interface with the Chargify API using simple ActiveRecord-like syntax, i.e.:

```ruby
Chargify::Subscription.create(
  :customer_reference => 'moklett',
  :product_handle => 'chargify-api-ares-test',
  :credit_card_attributes => {
    :first_name => "Michael",
    :last_name => "Klett",
    :expiration_month => 1,
    :expiration_year => 2010,
    :full_number => "1234-1234-1234-1234"
  }
)

subscription.credit_card_attributes = { :expiration_year => 2013 }
subscription.save

subscription.cancel
```

##### Note

Updating nested resources is _not_ supported nor recommended. If you wish to update a subscriptions customer please do so by updating the customer object itself.

Bad:

```ruby
subscription = Chargify::Subscription.find(123)
subscription.customer.first_name = 'fred'
subscription.customer.save
```

Good:

```ruby
subscription = Chargify::Subscription.find(123)
customer     = Chargify::Customer.find(subscription.customer.id)
customer.first_name = 'fred'
customer.save
```

Check out the examples in the `examples` directory. If you're not familiar with how ActiveResource works, you may be interested in some [ActiveResource Documentation](http://apidock.com/rails/ActiveResource/Base)

### Duplicate Prevention

[Duplicate Prevention](https://docs.chargify.com/api-duplicate-prevention) is supported by including the `uniqueness_token` attribute, for example:

```ruby
subscription = Chargify::Subscription.create(
  :product_handle => 'product-with-trial',
  :customer_attributes => {
    :first_name => "Test",
    :last_name => "Customer",
    :email => "test@example.com"
  },
  :uniqueness_token => "abc-123-def-456"
)
```

### Statements PDF retrieving

```ruby
File.open(file_path, 'wb+') do |f|
  f.write Chargify::Statement.find_pdf(statement.id)
end
```

### Compatibility

- Rails/ActiveResource 2.3.x, use 0.5.x
- Rails/ActiveResource 3.x, use 0.6 and up

| chargify_api_ares | Rails 2.3.x  | Rails 3.0.0 - 3.0.19 | Rails 3.0.20 and up |
| ----------------- | ------------ | -------------------- | ------------------- |
| 0.5.x             | OK           | Incompatible         | OK                  |
| 0.6.x             | Incompatible | OK (Monkey-patched)  | OK                  |

#### The problem with Rails/ActiveResource/ActiveModel 3.0.0 - 3.0.19

Prior to Feb 12, 2013, Chargify would silently refuse to parse XML which
contained data specified as YAML, such as:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<person>
  <name>John</name>
  <email type="yaml" nil="true"/></email>
</person>
```

After Feb 12, 2013, Chargify returns a `400 Bad Request` response if
your XML contains any `type="yaml"` attribute, since there is no valid
reason to send YAML serialized data to Chargify and doing so smells
strongly of an attempt to exploit
[CVE-2013-0156](https://groups.google.com/forum/?fromgroups=#!topic/rubyonrails-security/61bkgvnSGTQ).

However, Rails/ActiveModel versions 3.0.0 to 3.0.19 had a bug (see
<https://github.com/rails/rails/pull/8853>) where any nil attribute
would have a `type="yaml"` attribute added during XML serialization.

Using ActiveResource 3.0.0 - 3.0.19 along with 0.5.x or lower of this
gem may result in your sending `type="yaml"` XML to Chargify. Thus, your
requests will be rejected.

Version 0.6.x of this gem will attempt to patch your ActiveModel if you
have an incompatible version. To avoid this patch, you should use
3.0.20 or higher of ActiveResource.

### Contributing

- Check out the latest master to make sure the feature hasn't been
  implemented or the bug hasn't been fixed yet
- Check out the [issue
  tracker](http://github.com/chargify/chargify_api_ares/issues) to make
  sure someone already hasn't requested it and/or contributed it
- Fork the project
- Start a feature/bugfix branch
- Commit and push until you are happy with your contribution
- Make sure to add tests for the feature/bugfix
- Please squash your commits.
- Please try not to mess with the Rakefile, version, or history. If you
  want to have your own version, or is otherwise necessary, that is
  fine, but please isolate it to its own commit so we can cherry-pick
  around it.

### Contributors

We would like to publicly thank [everyone who has contributed](https://github.com/chargify/chargify_api_ares/graphs/contributors) their time
and talents to developing this gem. Thank you!


HISTORY.md
## 1.4.15 / June 20, 2018

- Changes Subscription#remove_coupon to use coupon_code instead of code [PR 165](https://github.com/chargify/chargify_api_ares/pull/165).  
  With this change, the currently applied coupon code must be passed for removal to occur.

## 1.4.14 / December 7, 2017

- Allows canceling a subscription with all supported attributes [PR 161](https://github.com/chargify/chargify_api_ares/pull/161)

## 1.4.13 / July 21, 2017

- Note: No changes from 1.4.12, previous release was not officially published
- Adds ability to get ReasonCodes [PR 154](https://github.com/chargify/chargify_api_ares/pull/154)

## 1.4.12 / July 18, 2017

- Adds ability to get ReasonCodes [PR 154](https://github.com/chargify/chargify_api_ares/pull/154)

## 1.4.11 / May 12, 2017

- Add Subscription Retry [PR 155](https://github.com/chargify/chargify_api_ares/pull/155) and [PR 145](https://github.com/chargify/chargify_api_ares/pull/145) by @gotchahn

## 1.4.7 / August 22 2016

- Adding Invoice Payment creation [PR 138] (https://github.com/chargify/chargify_api_ares/pull/138) by @carsonreinke

## 1.4.6 / August 10 2016

- Allow filtering Invoices by status [PR 139] (https://github.com/chargify/chargify_api_ares/pull/139) by @speric

## 1.4.5 / July 14 2016

- Adds ability to find pdfs from invoices, show subscription invoices [PR 135] (https://github.com/chargify/chargify_api_ares/pull/135)

## 1.4.4 / Jun 3 2016

- Adds support for External Payments [PR 128](https://github.com/chargify/chargify_api_ares/pull/128)

## 1.4.3 / Jan 5 2016

- Force json format for bulk allocations endpoints [PR 126](https://github.com/chargify/chargify_api_ares/pull/126) by @ryansch

## 1.4.2 / Jan 4 2016

- Adds `Chargify::Allocations.bulk_create` to work with https://docs.chargify.com/api-allocations#create-multiple-allocations. [PR 125](https://github.com/chargify/chargify_api_ares/pull/125) by @ryansch

## 1.4.1 / Nov 12 2015

- Adds `paypal_account` nested attribute to subscription [PR 119](https://github.com/chargify/chargify_api_ares/pull/119) by @richmisen

## 1.4.0 / Oct 6 2015

- Reverts custom `load_remote_errors` for Migration API (https://github.com/chargify/chargify_api_ares/pull/118)
- Adds a `Chargify::Statement.find_pdf` method (https://github.com/chargify/chargify_api_ares/pull/116)

## 1.3.5 / Aug 12 2015

- Adds support for customer metadata

## 1.3.4 / May 14 2015

- Adds duplicate prevention support

## 1.3.3 / Mar 27 2015

- When reconfiguring, the site should recalculate based on new settings

## 1.3.2 / Dec 15 2014

- `product_change` and `cancel_delayed_product_change` added to Subscription
  resource

## 1.3.1 / Dec 15 2014

- Update metadata / metafield endpoints to work with activeresource 4+

## 1.1.0 / Aug 20 2014

- Added cancellation message option for subscription canceling
- restore 1.8.7 style hash syntax
- Add metafields and metadata
- Remove 'bank_account' attribute from subscription upon save
- Update Gemfile source to https://rubygems.org due to Bundler deprecation
- Update payment_profile on subscriptions to return credit_card or bank_account

## 1.0.5 / May 11 2014

### Backwards-incompatible changes

- `Chargify::Subscription` methods no longer raise exception when there is a validation error. Now you must inspect the subscription object for errors. eg:

```ruby
subscription.reactivate

if subscription.errors.any?
  # handle errors
end
```

## 1.0.0 / Nov 19 2013

### Backwards-incompatible changes

- `Chargify::Subscription.charge` now returns an ActiveResource `Charge` object. In the case of an error, the `Charge` object will have `errors`, and you will not have to rescue an HTTP `422`.

- Adds new `Chargify::Migration` and `Chargify::Migration::Preview` resources. These can be used as follows:

```ruby
subscription = Chargify::Subscription.find_by_customer_reference('marky-mark')

# Chargify::Migration
migration = subscription.migrate(:product_handle => "basic-plan")

migration = Chargify::Migration.create(:subscription_id => subscription.id, :product_handle => "basic-plan")

# Chargify::Migration::Preview
preview = Chargify::Migration::Preview.create(:subscription_id => subscription.id, :product_handle => "basic-plan")

preview = Chargify::Migration.preview(:subscription_id => subscription.id, :product_handle => "basic-plan")
```

Error handling looks like:

```ruby
migration = subscription.migrate(:product_handle => "non-existent-plan")
migration.errors.full_messages
# => ["Invalid Product"]

preview = Chargify::Migration.preview(:subscription_id => subscription.id, :product_handle => "non-existent-plan")
preview.errors.full_messages
# => ["Product must be specified"]
```

See `examples/migrations.rb` and specs for more details.


examples/charges.rb
$: << File.expand_path(File.dirname(__FILE__) + '/../lib')

require 'chargify_api_ares'

# You could load your credentials from a file...
chargify_config = YAML::load_file(File.join(File.dirname(__FILE__), '..', 'chargify.yml'))

Chargify.configure do |c|
  c.subdomain = chargify_config['subdomain']
  c.api_key   = chargify_config['api_key']
end

# Lookup up existing Subscription using the Customer's reference
subscription = Subscription.find_by_customer_reference('moklett')
# => #<Chargify::Subscription:0x1020ed4b0 @prefix_options={}, @attributes={"cancellation_message"=>nil, "activated_at"=>Tue Nov 17 16:00:17 UTC 2009, "expires_at"=>nil, "updated_at"=>Tue Nov 17 16:00:17 UTC 2009, "credit_card"=>#<Chargify::Subscription::CreditCard:0x102046b10 @prefix_options={}, @attributes={"card_type"=>"bogus", "expiration_year"=>2020, "masked_card_number"=>"XXXX-XXXX-XXXX-1", "first_name"=>"Michael", "expiration_month"=>1, "last_name"=>"Klett"}>, "product"=>#<Chargify::Product:0x10204a2d8 @prefix_options={}, @attributes={"price_in_cents"=>0, "name"=>"Chargify API Ares Test", "handle"=>"chargify-api-ares-test", "product_family"=>#<Chargify::Product::ProductFamily:0x1020490b8 @prefix_options={}, @attributes={"name"=>"Chargify API ARes Test", "handle"=>"chargify-api-ares-test", "id"=>79, "accounting_code"=>nil}>, "id"=>153, "accounting_code"=>nil, "interval_unit"=>"month", "interval"=>1}>, "trial_ended_at"=>nil, "id"=>331, "current_period_ends_at"=>Thu Dec 17 16:00:17 UTC 2009, "product_handle"=>"chargify-api-ares-test", "trial_started_at"=>nil, "customer"=>#<Chargify::Customer:0x10204b688 @prefix_options={}, @attributes={"reference"=>"moklett", "updated_at"=>Tue Nov 17 15:51:02 UTC 2009, "id"=>331, "first_name"=>"Michael", "organization"=>"Chargify", "last_name"=>"Klett", "email"=>"moklett@example.com", "created_at"=>Tue Nov 17 15:51:02 UTC 2009}>, "balance_in_cents"=>0, "current_period_started_at"=>Tue Nov 17 16:00:17 UTC 2009, "state"=>"active", "created_at"=>Tue Nov 17 16:00:17 UTC 2009, "customer_reference"=>"moklett"}>

# Perform a one-time charge against this subscription
Chargify::Charge.create(:subscription_id => subscription.id, :amount => 10.00, :memo => 'Extra service')
# => #<Chargify::Charge:0x007f80da97b6b8 @attributes={"amount"=>10.00, "memo"=>"Extra service", "id"=>43382, "amount_in_cents"=>1000, "component_id"=>nil, "created_at"=>2013-11-08 21:16:03 UTC, "ending_balance_in_cents"=>1000, "kind"=>"one_time", "payment_id"=>4383, "product_id"=>153, "starting_balance_in_cents"=>0, "success"=>true, "tax_id"=>nil, "type"=>"Charge", "transaction_type"=>"charge", "gateway_transaction_id"=>nil, "statement_id"=>25695, "customer_id"=>40210}, @prefix_options={:subscription_id=>416940}, @persisted=true, @remote_errors=nil, @validation_context=nil, @errors=#<ActiveResource::Errors:0x007f80da97b0f0 @base=#<Chargify::Charge:0x007f80da97b6b8 ...>, @messages={}>>

# If a charge fails
charge = Chargify::Charge.create(:subscription_id => subscription.id, :amount => 10.00, :memo =>'Extra service')
# => #<Chargify::Charge:0x007f80dcc9f428 @attributes={"amount"=>10.0, "memo"=>"Extra service"}, @prefix_options={:subscription_id=>41630}, @persisted=false, @remote_errors=#<ActiveResource::ResourceInvalid: Failed.  Response code = 422.  Response message = Unprocessable Entity.>, @validation_context=nil, @errors=#<ActiveResource::Errors:0x007f80dcc9f0b8 @base=#<Chargify::Charge:0x007f80dcc9f428 ...>, @messages={:base=>["Bogus Gateway: Forced failure"]}>>

charge.valid?
# => false

charge.errors.full_messages
# => ["Bogus Gateway: Forced failure"]

# Or call 'charge' right on the subscription
subscription.charge(:amount => 10.00, :memo => 'Extra service')
# => #<Chargify::Charge:0x007f80da97b6b8 @attributes={"amount"=>10.00, "memo"=>"Extra service", "id"=>4313282, "amount_in_cents"=>1000, "component_id"=>nil, "created_at"=>2013-11-08 21:16:03 UTC, "ending_balance_in_cents"=>1000, "kind"=>"one_time", "payment_id"=>43303, "product_id"=>153, "starting_balance_in_cents"=>0, "success"=>true, "tax_id"=>nil, "type"=>"Charge", "transaction_type"=>"charge", "gateway_transaction_id"=>nil, "statement_id"=>25695, "customer_id"=>48210}, @prefix_options={:subscription_id=>41690}, @persisted=true, @remote_errors=nil, @validation_context=nil, @errors=#<ActiveResource::Errors:0x007f80da97b0f0 @base=#<Chargify::Charge:0x007f80da97b6b8 ...>, @messages={}>>
